omnigibson/prims/xform_prim.py

```diff
@@ -168,53 +169,114 @@ def has_material(self):
         material_path = self._binding_api.GetDirectBinding().GetMaterialPath().pathString
         return material_path != ""
 
-    def set_position_orientation(self, position=None, orientation=None):
+    def set_position_orientation(
+        self, position=None, orientation=None, frame: Literal["world", "parent", "scene"] = "world"
+    ):
         """
-        Sets prim's pose with respect to the world frame
+        Sets prim's pose with respect to the specified frame
 
         Args:
             position (None or 3-array): if specified, (x,y,z) position in the world frame
                 Default is None, which means left unchanged.
             orientation (None or 4-array): if specified, (x,y,z,w) quaternion orientation in the world frame.
                 Default is None, which means left unchanged.
+            frame (Literal): frame to set the pose with respect to, defaults to "world". parent frame
+            set position relative to the object parent. scene frame set position relative to the scene.
         """
-        if position is None or orientation is None:
-            current_position, current_orientation = self.get_position_orientation()
-            position = current_position if position is None else position
-            orientation = current_orientation if orientation is None else orientation
 
-        position = position if isinstance(position, th.Tensor) else th.tensor(position, dtype=th.float32)
-        orientation = orientation if isinstance(orientation, th.Tensor) else th.tensor(orientation, dtype=th.float32)
-        assert math.isclose(
-            th.norm(orientation).item(), 1, abs_tol=1e-3
-        ), f"{self.prim_path} desired orientation {orientation} is not a unit quaternion."
+        assert frame in ["world", "parent", "scene"], f"Invalid frame '{frame}'. Must be 'world', 'parent', or 'scene'."
+
+        # get the desired position and orientation in the specified frame
+        position, orientation = T.compute_desired_pose_in_frame(self, position, orientation, frame=frame)
+
+        if frame == "world" or frame == "scene":
+
```

Comment:
Remove the empty line. Remove the compute_desired_... call above. Move the assertion below to be before the if/else block. Then move the parent-local code block below (from my_world_transform until assert) into a `convert_pose_to_local` function in PoseAPI. Then convert this if/else a clean if/elif/else as described in entityprim.

Reply:
same thing here, the scene transform happens before the if statement.